<h1 align="center"><font size="5">Final Project: Rain Prediction in Australia with Machine Learning and Classification Models</font></h1>
<h1 align="center"><font size="3">This project was done as a part of Honors Content in Machine Learning with Python Couse in the IBM Data Science Professional Certificate.</font></h1>

<h1 align="center"><font size="2">by Sara Mezuri</font></h1>


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
    <li><a href="https://#Section_1">Instructions</a></li>
    <li><a href="https://#Section_2">About the Data</a></li>
    <li><a href="https://#Section_3">Importing Data </a></li>
    <li><a href="https://#Section_4">Data Preprocessing</a> </li>
    <li><a href="https://#Section_5">One Hot Encoding </a></li>
    <li><a href="https://#Section_6">Train and Test Data Split </a></li>
    <li><a href="https://#Section_7">Train Logistic Regression, KNN, Decision Tree, SVM, and Linear Regression models and return their appropriate accuracy scores</a></li>
</a></li>

<hr>


# Instructions


In this notebook, we will explore various classification algorithms to predict whether it will rain tomorrow in Australia using historical weather data. We will utilize the following machine learning algorithms taught in this course:

1. Linear Regression
2. KNN
3. Decision Trees
4. Logistic Regression
5. SVM

To evaluate the performance of these models, we will use several metrics, including:

1.  Accuracy Score
2.  Jaccard Index
3.  F1-Score
4.  LogLoss
5.  Mean Absolute Error
6.  Mean Squared Error
7.  R2-Score

Finally, we will generate a report summarizing the findings and performance of the models.

# About The Dataset


The original source of the data is Australian Government's Bureau of Meteorology and the latest data can be gathered from [http://www.bom.gov.au/climate/dwo/](http://www.bom.gov.au/climate/dwo/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01).

The dataset to be used has extra columns like 'RainToday' and our target is 'RainTomorrow', which was gathered from the Rattle at [https://bitbucket.org/kayontoga/rattle/src/master/data/weatherAUS.RData](https://bitbucket.org/kayontoga/rattle/src/master/data/weatherAUS.RData?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01)




This dataset contains observations of weather metrics for each day from 2008 to 2017. The **weatherAUS.csv** dataset includes the following fields:

| Field         | Description                                           | Unit            | Type   |
| ------------- | ----------------------------------------------------- | --------------- | ------ |
| Date          | Date of the Observation in YYYY-MM-DD                 | Date            | object |
| Location      | Location of the Observation                           | Location        | object |
| MinTemp       | Minimum temperature                                   | Celsius         | float  |
| MaxTemp       | Maximum temperature                                   | Celsius         | float  |
| Rainfall      | Amount of rainfall                                    | Millimeters     | float  |
| Evaporation   | Amount of evaporation                                 | Millimeters     | float  |
| Sunshine      | Amount of bright sunshine                             | hours           | float  |
| WindGustDir   | Direction of the strongest gust                       | Compass Points  | object |
| WindGustSpeed | Speed of the strongest gust                           | Kilometers/Hour | object |
| WindDir9am    | Wind direction averaged of 10 minutes prior to 9am    | Compass Points  | object |
| WindDir3pm    | Wind direction averaged of 10 minutes prior to 3pm    | Compass Points  | object |
| WindSpeed9am  | Wind speed averaged of 10 minutes prior to 9am        | Kilometers/Hour | float  |
| WindSpeed3pm  | Wind speed averaged of 10 minutes prior to 3pm        | Kilometers/Hour | float  |
| Humidity9am   | Humidity at 9am                                       | Percent         | float  |
| Humidity3pm   | Humidity at 3pm                                       | Percent         | float  |
| Pressure9am   | Atmospheric pressure reduced to mean sea level at 9am | Hectopascal     | float  |
| Pressure3pm   | Atmospheric pressure reduced to mean sea level at 3pm | Hectopascal     | float  |
| Cloud9am      | Fraction of the sky obscured by cloud at 9am          | Eights          | float  |
| Cloud3pm      | Fraction of the sky obscured by cloud at 3pm          | Eights          | float  |
| Temp9am       | Temperature at 9am                                    | Celsius         | float  |
| Temp3pm       | Temperature at 3pm                                    | Celsius         | float  |
| RainToday     | If there was rain today                               | Yes/No          | object |
| RainTomorrow  | If there is rain tomorrow                             | Yes/No          | float  |

Column definitions were gathered from [http://www.bom.gov.au/climate/dwo/IDCJDW0000.shtml](http://www.bom.gov.au/climate/dwo/IDCJDW0000.shtml?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01)



## **Import the required libraries**


First, we need to import the necessary libraries for data processing, visualization, and model building. If your environment does not support `!mamba install`, use `!pip install`.

In [1]:
# Surpress warnings:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn

In [3]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn import preprocessing
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm
from sklearn.metrics import jaccard_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
from sklearn.metrics import confusion_matrix, accuracy_score
import sklearn.metrics as metrics

### Importing the Dataset


Next, we load the dataset into a pandas DataFrame for exploration and preprocessing.

In [ ]:
# from pyodide.http import pyfetch

# async def download(url, filename):
#   response = await pyfetch(url)
#   if response.status == 200:
#       with open(filename, "wb") as f:
#           f.write(await response.bytes())

In [ ]:
# path='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillUp/labs/ML-FinalAssignment/Weather_Data.csv'

In [ ]:
# await download(path, "Weather_Data.csv")
# filename ="Weather_Data.csv"

In [ ]:
# df = pd.read_csv("Weather_Data.csv")

> Note: This version of the lab is designed for JupyterLite, which necessitates downloading the dataset to the interface. However, when working with the downloaded version of this notebook on your local machines (Jupyter Anaconda), you can simply **skip the steps above of "Importing the Dataset"** and use the URL directly in the `pandas.read_csv()` function. 


In [5]:
filepath = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ML0101EN-SkillUp/labs/ML-FinalAssignment/Weather_Data.csv"
df = pd.read_csv(filepath)

In [7]:
df.head()

Date  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine WindGustDir  \
0  2/1/2008     19.5     22.4      15.6          6.2       0.0           W   
1  2/2/2008     19.5     25.6       6.0          3.4       2.7           W   
2  2/3/2008     21.6     24.5       6.6          2.4       0.1           W   
3  2/4/2008     20.2     22.8      18.8          2.2       0.0           W   
4  2/5/2008     19.7     25.7      77.4          4.8       0.0           W   

   WindGustSpeed WindDir9am WindDir3pm  ...  Humidity9am  Humidity3pm  \
0             41          S        SSW  ...           92           84   
1             41          W          E  ...           83           73   
2             41        ESE        ESE  ...           88           86   
3             41        NNE          E  ...           83           90   
4             41        NNE          W  ...           88           74   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1017.6       1017.4         8         8     20.7     20.9        Yes   
1       1017.9       1016.4         7         7     22.4     24.8        Yes   
2       1016.7       1015.6         7         8     23.5     23.0        Yes   
3       1014.2       1011.8         8         8     21.4     20.9        Yes   
4       1008.3       1004.8         8         8     22.5     25.5        Yes   

   RainTomorrow  
0           Yes  
1           Yes  
2           Yes  
3           Yes  
4           Yes  

[5 rows x 22 columns]

### Data Preprocessing


#### One Hot Encoding


To prepare the data for machine learning algorithms, we need to convert categorical variables into binary variables using one hot encoding.


In [9]:
df_sydney_processed = pd.get_dummies(data=df, columns=['RainToday', 'WindGustDir', 'WindDir9am', 'WindDir3pm'])

Next, we replace the values of the 'RainTomorrow' column changing them from a categorical column to a binary column. We do not use the `get_dummies` method because we would end up with two columns for 'RainTomorrow' and we do not want, since 'RainTomorrow' is our target.


In [12]:
df_sydney_processed.replace(['No', 'Yes'], [0,1], inplace=True)

### Training Data and Test Data


Train/Test Split involves splitting the dataset into training and testing sets respectively, which are mutually exclusive. After which, you train with the training set and test with the testing set. This will provide a more accurate evaluation on out-of-sample accuracy because the testing dataset is not part of the dataset that have been used to train the model. Therefore, it gives us a better understanding of how well our model generalizes on new data. 

Now, we set our `features` or x values and our Y or target variable. 


In [16]:
df_sydney_processed.drop('Date',axis=1,inplace=True)

In [18]:
df_sydney_processed = df_sydney_processed.astype(float)

In [20]:
features = df_sydney_processed.drop(columns='RainTomorrow', axis=1)
Y = df_sydney_processed['RainTomorrow']

### Linear Regression

Linear Regression fits a linear model with coefficients B = (B1, ..., Bn) to minimize the 'residual sum of squares' between the actual value y in the dataset, and the predicted value yhat using linear approximation.

### Model Training

We use the `train_test_split` function to split the `features` and `Y` dataframes with a `test_size` of `0.2` and the `random_state` set to `10`. This means that 80% of the data used for training and 20% for testing. The random_state is set to 10 to ensure reproducibility.

In [22]:
# Split the features and Y dataframes
x_train, x_test, y_train, y_test = train_test_split(features, Y, test_size=0.2, random_state=10)

# Print the shape of x_train and y_train to ensure that the dimensions match.
print(x_train.shape)
print(y_train.shape)

# Print the shape of x_test and y_test to ensure that the dimensions match.
print(x_test.shape)
print(y_test.shape)

(2616, 66)
(2616,)
(655, 66)
(655,)


### Making Predictions

We create and train a Linear Regression model called LinearReg using the training data (`x_train`, `y_train`).


In [24]:
from sklearn import linear_model

# Create and train a Linear Regression model
LinearReg = linear_model.LinearRegression()
LinearReg.fit (x_train, y_train)

# Print the coefficients of the trained model
print ('Coefficients: ', LinearReg.coef_)

Coefficients:  [-2.36946709e-02  1.30528498e-02  7.30741811e-04  6.48086479e-03
 -3.51665228e-02  4.23576504e-03  1.83177620e-03  7.91415854e-04
  9.59311783e-04  8.55637076e-03  7.70190836e-03 -9.24845557e-03
 -8.86878011e-03  1.00413073e-02  1.44858937e-02 -3.54592115e-03
 -7.74520021e+10 -7.74520021e+10  2.11152148e+09  2.11152148e+09
  2.11152148e+09  2.11152148e+09  2.11152148e+09  2.11152148e+09
  2.11152148e+09  2.11152148e+09  2.11152148e+09  2.11152148e+09
  2.11152148e+09  2.11152148e+09  2.11152148e+09  2.11152148e+09
  2.11152148e+09  2.11152148e+09  7.65756278e+09  7.65756278e+09
  7.65756278e+09  7.65756278e+09  7.65756278e+09  7.65756278e+09
  7.65756278e+09  7.65756278e+09  7.65756278e+09  7.65756278e+09
  7.65756278e+09  7.65756278e+09  7.65756278e+09  7.65756278e+09
  7.65756278e+09  7.65756278e+09 -9.48859411e+09 -9.48859411e+09
 -9.48859411e+09 -9.48859411e+09 -9.48859411e+09 -9.48859411e+09
 -9.48859411e+09 -9.48859411e+09 -9.48859411e+09 -9.48859411e+09
 -9.488594

Now, we use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.

In [26]:
# Use the predict method on the testing data
predictions = LinearReg.predict(x_test)

x = np.asanyarray(x_test)
y = np.asanyarray(y_test)

### Evaluation

We compare the actual values and predicted values to evaluate the model's performance. Evaluation metrics provide insights into areas that require improvement. The metrics used are:

**Mean Absolute Error (MAE)**: The mean of the absolute value of the errors.

**Mean Squared Error (MSE)**: The mean of the squared errors, which gives more weight to larger errors.

**R-squared (R2 Score)**: Indicates how close the data points are to the fitted regression line. A higher value indicates a better fit. The best possible score is 1.0 and it can be negative (because the model can be arbitrarily worse).

Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [28]:
from sklearn.metrics import r2_score

# Calculate the metrics
LinearRegression_MAE = np.mean(np.absolute(predictions - y_test))
LinearRegression_MSE = np.mean((predictions - y_test) ** 2)
LinearRegression_R2 = r2_score(y_test , predictions)

# Print the values for each metric
print(LinearRegression_MAE)
print(LinearRegression_MSE)
print(LinearRegression_R2)

0.2563250126729485
0.1157232929703616
0.4271186909603826


We show the MAE, MSE, and R2 in a tabular format using data frame for the linear model.


In [30]:
# Create a DataFrame to store the metrics
Report = pd.DataFrame({'Metric': ['Mean Absolute Error', 'Mean Squared Error', 'R-squared (R2 Score)'],
                           'Value': [LinearRegression_MAE, LinearRegression_MSE, LinearRegression_R2]})

# Display the DataFrame
Report

Metric     Value
0   Mean Absolute Error  0.256325
1    Mean Squared Error  0.115723
2  R-squared (R2 Score)  0.427119

These metrics indicate that the Linear Regression model has moderate predictive performance, with an R-squared value suggesting that approximately 42.7% of the variance in the target variable is explained by the model.

### KNN


**K-Nearest Neighbors (KNN)** is a supervised learning algorithm that is used for classification and regression tasks. In classification, the algorithm 'trains' on data points corresponding to their classifications. To predict the class of a given data point, KNN considers the classes of the 'K' nearest data points and chooses the class that is most common among them as the predicted class.

### Model Training

We will create and train a KNN model called KNN using the training data (`x_train`, `y_train`) with the `n_neighbors` parameter set to `4`.

In [32]:
# Set the parameter k to 4
k = 4

# Creat and train the KNN model 
KNN = KNeighborsClassifier(n_neighbors = k).fit(x_train,y_train)
KNN

KNeighborsClassifier(n_neighbors=4)

### Making Predictions

We will use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.

In [35]:
# Predict on the test data
KNN_predictions = KNN.predict(x_test)

# Display the first 10 predictions
KNN_predictions[0:10]

array([0., 0., 1., 0., 0., 0., 0., 1., 1., 0.])

### Evaluation

In multilabel classification, **accuracy classification score** is a function that computes subset accuracy. It calculates how closely the actual labels and predicted labels match in the test set.

The **Jaccard Index** is the size of the intersection divided by the size of the union of the two label sets. If the entire set of predicted labels for a sample strictly matches the true set of labels, then the subset accuracy is 1.0; otherwise, it is 0.0.

And **F1 Score** is given as the harmonic mean of precision and recall, providing a balance between the two.

Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.


In [38]:
# Calculate the Accuracy Score
KNN_Accuracy_Score = metrics.accuracy_score(y_test, KNN_predictions)
print("KNN_Accuracy_Score = ", KNN_Accuracy_Score)

# Calculate the Jaccard Index
KNN_JaccardIndex = metrics.jaccard_score(y_test, KNN_predictions)
print("KNN_JaccardIndex = ", KNN_JaccardIndex)

# Calculate the F1 Score
KNN_F1_Score = metrics.f1_score(y_test, KNN_predictions)
print("KNN_F1_Score = ", KNN_F1_Score)


KNN_Accuracy_Score =  0.8183206106870229
KNN_JaccardIndex =  0.4251207729468599
KNN_F1_Score =  0.5966101694915255


These metrics indicate that the KNN model has a good balance of precision and recall, and an accuracy of approximately 82%.

### Decision Tree


A **Decision Tree** is a supervised learning algorithm that can be used for both classification and regression tasks. The goal is to create a model that predicts the value of a target variable by learning simple decision rules inferred from the data features. Decision trees are intuitive and easy to interpret.

### Model Training

We will create and train a Decision Tree model called Tree using the training data (`x_train`, `y_train`). We'll set the criterion to "entropy" and the maximum depth of the tree to 4.

In [43]:
# Create the Decision Tree model
Tree = DecisionTreeClassifier(criterion="entropy", max_depth = 4)
Tree # the default parameters

# Train the model
Tree.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=4)

### Making Predictions

Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.

In [46]:
# Predict on the test data
T_predictions = Tree.predict(x_test)

print (T_predictions[0:10])
print (y_test[0:10])

[0. 0. 1. 0. 0. 0. 0. 1. 1. 0.]
3066    0.0
2456    0.0
1629    1.0
838     1.0
2506    0.0
130     1.0
2404    1.0
1226    1.0
290     1.0
2136    0.0
Name: RainTomorrow, dtype: float64


### Evaluation

Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.

In [49]:
# Calculate the Accuracy Score
Tree_Accuracy_Score = metrics.accuracy_score(y_test, T_predictions)
print(" Tree_Accuracy_Score = ", Tree_Accuracy_Score )

# Calculate the Jaccard Index
Tree_JaccardIndex = metrics.jaccard_score(y_test, T_predictions)
print(" Tree_JaccardIndex = ", Tree_JaccardIndex )

# Calculate the F1 Score
Tree_F1_Score = metrics.f1_score(y_test, T_predictions)
print(" Tree_F1_Score = ", Tree_F1_Score )


 Tree_Accuracy_Score =  0.8183206106870229
 Tree_JaccardIndex =  0.48034934497816595
 Tree_F1_Score =  0.6489675516224189


These metrics indicate that the Decision Tree model performs well, with an accuracy of approximately 82% and a balanced precision and recall as indicated by the F1 Score.

### Logistic Regression


**Logistic Regression** is a supervised learning algorithm used for binary classification tasks. Unlike Linear Regression, which predicts continuous values, Logistic Regression predicts the probability of a categorical dependent variable. The output is typically a probability that indicates the likelihood of a particular class.

### Model Training

Use the `train_test_split` function to split the `features` and `Y` dataframes with a `test_size` of `0.2` and the `random_state` set to `1`.

In [54]:
# Split the features and Y dataframes
x_train, x_test, y_train, y_test = train_test_split( features, Y, test_size=0.2, random_state=1)

print ('Train set:', x_train.shape,  y_train.shape)
print ('Test set:', x_test.shape,  y_test.shape)

Train set: (2616, 66) (2616,)
Test set: (655, 66) (655,)


Create and train a LogisticRegression model called LR using the training data (`x_train`, `y_train`) with the `solver` parameter set to `liblinear`.

In [57]:
# Create and train the Logistic Regression model
LR = LogisticRegression(solver='liblinear')
LR.fit(x_train,y_train)

LogisticRegression(solver='liblinear')

### Making Predictions

Now, use the `predict` and `predict_proba` methods on the testing data (`x_test`) and save it as 2 arrays `predictions` and `predict_proba`.

In [60]:
# Use the predict method to predict on the test data
LR_predictions =  LR.predict(x_test)

print("LR Predictions:")
print(LR_predictions)

LR Predictions:
[0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0. 1. 0. 0. 0. 1. 1. 0. 0.
 0. 0. 0. 0. 0. 0. 1. 0. 0. 1. 1. 1. 1. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 1. 0. 1. 0. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 1. 0. 0. 1. 0. 0. 1. 0.
 0. 0. 0. 0. 0. 1. 0. 1. 1. 0. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0.
 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 1. 0. 0.
 0. 1. 0. 1. 1. 0. 0. 0. 1. 0. 0. 1. 0. 0. 1. 0. 1. 0. 1. 1. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 1. 1. 0. 0. 0. 0. 0. 0.
 1. 0. 0. 1. 0. 1. 0. 0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 1. 0. 0. 0. 1. 0. 0.
 0. 1. 0. 0. 0. 1. 0. 1. 1. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0. 0. 0. 0. 0. 1.
 0. 0. 1. 0. 0. 0. 1. 1. 1. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 1. 1. 1. 0. 0. 0.
 0. 1. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 0. 0. 0. 1. 0.
 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0

In [62]:
# Use the predict_proba method to predict on the test data
predict_proba = LR.predict_proba(x_test)

print("Predicted Probabilities:")
print(predict_proba)

Predicted Probabilities:
[[0.74087112 0.25912888]
 [0.97444658 0.02555342]
 [0.52599984 0.47400016]
 ...
 [0.97879482 0.02120518]
 [0.70338555 0.29661445]
 [0.21208177 0.78791823]]


### Evaluation

In logistic regression, the output can be the probability of customer churn is yes (or equals to 1). This probability is a value between 0 and 1. **Log loss( Logarithmic loss)** measures the performance of a classifier where the predicted output is a probability value between 0 and 1. 

Using the `predictions`, `predict_proba` and the `y_test` dataframe calculate the value for each metric using the appropriate function.

In [65]:
# Calculate the Accuracy Score
LR_Accuracy_Score =  metrics.accuracy_score(y_test, LR_predictions)
print(" LR_Accuracy_Score = ", LR_Accuracy_Score )

# Calculate the Jaccard Index
LR_JaccardIndex = jaccard_score(y_test, LR_predictions)
print(" LR_JaccardIndex = ", LR_JaccardIndex )

# Calculate the F1 Score
LR_F1_Score  = metrics.f1_score(y_test, LR_predictions)
print(" LR_F1_Score  = ", LR_F1_Score )

# Calculate the Log Loss
LR_Log_Loss = metrics.log_loss(y_test, predict_proba)
print(" LR_Log_Loss = ", LR_Log_Loss)

 LR_Accuracy_Score =  0.833587786259542
 LR_JaccardIndex =  0.5
 LR_F1_Score  =  0.6666666666666666
 LR_Log_Loss =  0.3811501324210304


These metrics indicate that the Logistic Regression model performs well in predicting the categorical outcomes with a high accuracy score and a reasonable F1 score. The log loss value also indicates a good fit for the probability predictions.

### SVM


**Support Vector Machine (SVM)** is a supervised learning algorithm used for classification tasks. It works by mapping data to a high-dimensional feature space, where a separator (hyperplane) between categories is found. This allows for categorization even when data is not linearly separable.

### Model Training

Create and train a SVM model called SVM using the training data (`x_train`, `y_train`).

In [70]:
# Create and train the SVM model
SVM = svm.SVC(kernel='rbf')
SVM.fit(x_train, y_train) 

SVC()

### Making Predictions 

Now use the `predict` method on the testing data (`x_test`) and save it to the array `predictions`.

In [72]:
# Predict on the test data
SVM_predictions = SVM.predict(x_test)
SVM_predictions[0:10]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

### Evaluation

Using the `predictions` and the `y_test` dataframe calculate the value for each metric using the appropriate function.

In [74]:
# Calculate the Accuracy Score
SVM_Accuracy_Score = metrics.accuracy_score(y_test, SVM_predictions)
print(" SVM_Accuracy_Score = ", SVM_Accuracy_Score )

# Calculate the Jaccard Index
SVM_JaccardIndex = jaccard_score(y_test, SVM_predictions, pos_label=0)
print(" SVM_JaccardIndex = ", SVM_JaccardIndex )

# Calculate the F1 Score
SVM_F1_Score  = metrics.f1_score(y_test, SVM_predictions,average='weighted')
print(" SVM_F1_Score = ", SVM_F1_Score )

 SVM_Accuracy_Score =  0.7221374045801526
 SVM_JaccardIndex =  0.7221374045801526
 SVM_F1_Score =  0.6056223268908018


These metrics indicate that the SVM model performs moderately well in predicting the categorical outcomes with a reasonable accuracy score and F1 score.

### Report


Show the Accuracy,Jaccard Index,F1-Score and LogLoss in a tabular format using data frame for all of the above models.

\*LogLoss is only for Logistic Regression Model


In [76]:
# Create a tabular format to store the evaluation metrics for each model
d_f = {
    'Model': ['KNN', 'Decision Tree', 'Logistic Regression', 'SVM'],
    'Accuracy': [KNN_Accuracy_Score, Tree_Accuracy_Score, LR_Accuracy_Score, SVM_Accuracy_Score],
    'Jaccard Index': [KNN_JaccardIndex, Tree_JaccardIndex, LR_JaccardIndex, SVM_JaccardIndex],
    'F1 Score': [KNN_F1_Score, Tree_F1_Score, LR_F1_Score, SVM_F1_Score],
    'Log Loss': [None, None, LR_Log_Loss, None]  # LogLoss is only for Logistic Regression
}

# Create the DataFrame 
Report = pd.DataFrame(d_f)

# Display the DataFrame
Report

Model  Accuracy  Jaccard Index  F1 Score  Log Loss
0                  KNN  0.818321       0.425121  0.596610       NaN
1        Decision Tree  0.818321       0.480349  0.648968       NaN
2  Logistic Regression  0.833588       0.500000  0.666667   0.38115
3                  SVM  0.722137       0.722137  0.605622       NaN

### Insights 

* Logistic Regression stands out with the highest accuracy, F1 Score, and has a reasonable Log Loss, making it the best-performing model for this dataset in terms of classification performance.

* SVM performs well in terms of Jaccard Index, indicating strong performance in predicting positive cases but has the lowest accuracy.

* KNN and Decision Tree have similar accuracy but differ in their balance of precision and recall. KNN, in particular, has lower performance in both Jaccard Index and F1 Score.

Each model has its strengths and weaknesses, and the choice of the best model can depend on specific requirements such as balancing precision and recall, overall accuracy, or the importance of probabilistic predictions.